### ISSUE TITLE
Cache-Control “private” and “no cache” are mutually exclusive

### ISSUE THREAD

**Issue by kinglozzer** [silverstripe/silverstripe-framework#8289]
A client noticed a few AJAX caching issues after upgrading a site to 3.7 - HTTP responses for `admin/pages/updatetreenodes?ids=123` after re-ordering pages were cached, so pages kept “popping” back to their old sort order after dragging & dropping.

The cache headers that were being set were:
```
Cache-Control: must-revalidate, private
Cache-Control: max-age=2592000
```

Maybe the max-age is the root cause of my client’s issue (their cPanel installation was adding that), and I resolved the issue by setting `HTTP.cache_ajax_requests: false`, but it struck me as odd that `no-cache` and `no-store` are removed inside the CMS.

---

After a bit more investigating, I found:

- Default `Cache-Control` header is `no-cache, no-store, must-revalidate`
- When a session cookie is present, that changes to `must-revalidate, private`

While adding `private` makes sense when a session cookie is present, I don’t see any reason for removing `no-cache` and `no-store`? `HTTPCacheControl::privateCache()` [explicitly removes](https://github.com/silverstripe/silverstripe-framework/blob/3.7.1/control/HTTPCacheControl.php#L361-L362) those two directives, and `HTTPCacheControl::disableCache()` will [remove _all_](https://github.com/silverstripe/silverstripe-framework/blob/3.7.1/control/HTTPCacheControl.php#L333-L337) existing directives, making it impossible to have both “private” and “no cache”.

HTTP caching is hard, as evidenced by the amount of discussion on the PR to implement this, so maybe there’s a good reason for this behaviour that I’m not aware of.

I’ve tested 4.2.0, but the `Cache-Control` headers don’t seem to change _at all_ when a session cookie is present... If I can confirm that’s not just a problem on my end, I’ll open a separate issue

**Comment by dhensby** [comment]
Yes, this is a critical bug that needs fixing in 3.7

**Comment by dhensby** [comment]
>Do you think this is related to the errors I have seen in my 3.7.1 logs as mentioned in this post:

Note that looks expected. You can't move down in priority without using the force option

**Comment by bcairns** [comment]
Can confirm we are seeing weird caching issues in the admin on multiple 3.7.x sites.

Typically the user is trying to edit a page or create a new one, and keeps getting sent to a previously visited page instead.

**Comment by kinglozzer** [comment]
Just been looking to recreate this and thought it’s worth mentioning for anyone else trying to recreate it - your environment type may be important. I can only recreate the issue in live mode.

**Comment by kinglozzer** [comment]
I think my original issue description may be wrong. While `privateCache()` and `disableCache()` may appear mutually exclusive, `disableCache()` takes priority so subsequent calls to `privateCache()` won’t actually have any effect.

The issue I was experiencing is specifically because the framework now caches AJAX requests by default. It appears that we used to _attempt_ to do that, but actually had the inverse effect:

1. https://github.com/silverstripe/silverstripe-framework/blob/3.6.6/control/HTTP.php#L345-L351 - If `cache_ajax_requests` is switched ON (the default), set `$cacheAge = 0;`
2. https://github.com/silverstripe/silverstripe-framework/blob/3.6.6/control/HTTP.php#L353-L397 - unless `$cacheAge` is greater than 0, set `no-cache`

Given the above, I would suggest we switch the default value for `HTTP.cache_ajax_requests` to `false` to restore the pre-3.7 behaviour.

**Comment by dhensby** [comment]
Are we seeing AJAX requests respond with 304s even when the content should be different? that's not so good :(

**Comment by kinglozzer** [comment]
@dhensby In my case the server is responding with a `200` status code, but the browser is still using the cached response:

<img width="533" alt="screen shot 2018-08-10 at 15 04 14" src="https://user-images.githubusercontent.com/1655548/43962110-bb92c200-9cae-11e8-830e-54f2af041afe.png">

The full response headers are:

```
Accept-Ranges: none
Cache-Control: max-age=2592000
Cache-Control: must-revalidate, private
Content-Type: text/json;charset=UTF-8
Date: Fri, 10 Aug 2018 14:02:34 GMT
Expires: Sun, 09 Sep 2018 14:02:34 GMT
Last-Modified: Fri, 10 Aug 2018 14:02:34 GMT
Vary: X-Requested-With
X-Controller: CMSPagesController
X-Frame-Options: SAMEORIGIN
X-Include-CSS: <a bunch of CSS>
X-Title: SilverStripe+-+Pages
X-UA-Compatible: IE=Edge,chrome=1
```

**Comment by kinglozzer** [comment]
Right, so I didn’t spot at first that the `Expires:` date above is `Sep` - i.e. +1 month. Turns out that comes from my `.htaccess`, even though I have the following directive:

```
ExpiresDefault                                      "access plus 1 month"
...
ExpiresByType application/json                      "access plus 0 seconds"
```

The above is having no effect because we’re using the wrong content type (`text/json`):

- https://github.com/silverstripe/silverstripe-framework/blob/3.7.1/admin/code/LeftAndMain.php#L1035
- https://github.com/silverstripe/silverstripe-framework/blob/3.7.1/control/PjaxResponseNegotiator.php#L96

So I think that’s the root cause (for me anyway). It just never happened before now because we always forced AJAX request caching off 😕

**Comment by dhensby** [comment]
OK - so I think this issue is going to be exacerbated by people running the default H5BP `.htaccess` file, particularly the removal of etags.

It should not be possible to get a cached response loaded by the browser because our responses all come with etags computed from the body of the response, so getting different responses on the same URL will be stored as separate cache entries.

@bcairns please can you provide some more indepth steps to reproduce the issues; I cannot reproduce any caching issues when adding pages. Are you using htlm 5 boilerplate `.htaccess` rules?

**Comment by chillu** [comment]
Forward ported to 4.x via https://github.com/silverstripe/silverstripe-admin/pull/629 and https://github.com/silverstripe/silverstripe-framework/pull/8360

**Comment by robbieaverill** [comment]
All PRs merged, feel free to reopen the issue if there's something that hasn't been resolved. @dhensby's original PR at #8318 has "fixes #8289", so in combination with that and @chillu's forward porting to 4.2 I'm happy to close
